### PR TITLE
Update params and add types to IP Range update API

### PR DESCRIPTION
**Pull request by mcvic1rj**
Added the following parameters:

- _id
- location
- overrideOrganization
- overrideLocation

Added types to new and existing parameters.

### CODE REVIEW

**Comment by PRMerger6** [comment]
@mcvic1rj : Thanks for your contribution! The author(s) have been notified to review your proposed change.

**Comment by ktoliver** [comment]
#label:"aq-pr-triaged"

File: CloudAppSecurityDocs/api-data-enrichment-update.md [hunk 1]
```diff
@@ -18,13 +18,17 @@ POST /api/v1/subnet/<ip_range_id>/update_rule/
...
+| Parameter | Type | Description |
+| --- | --- | --- |
+| name | string | The unique name of the range |
+| _id (Optional)| string | The IP range id. Required if you are not changing the name. |
+| category | int | The id of the range category. Providing a category helps you easily recognize activities from interesting IP addresses. Possible values include:<br /><br />**1**: Corporate<br />**2**: Administrative<br />**3**: Risky<br />**4**: VPN<br />**5**: Cloud provider<br />**6**: Other |
+| subnets | list | An array of masks as strings (IPv4 / IPv6) |
+| overrideOrganization (Optional) | bool | Flag to enable or disable overriding the Organization field. |
+| organization (Optional) | string | The registered ISP |
+| tags (Optional) | list | An array of new or existing objects including the tag name, id, description, name template, and tenant id |
+| overrideLocation (Optional) | bool | Flag to enable or disable overriding the Location field. |
```
Comment (dagoltz) [inline CloudAppSecurityDocs/api-data-enrichment-update.md:30]: remove the overrideLocation - it's not needed

File: CloudAppSecurityDocs/api-data-enrichment-update.md [hunk 2]
```diff
@@ -18,13 +18,17 @@ POST /api/v1/subnet/<ip_range_id>/update_rule/
...
-| --- | --- |
-| name | The unique name of the range |
-| category | The id of the range category. Providing a category helps you easily recognize activities from interesting IP addresses. Possible values include:<br /><br />**1**: Corporate<br />**2**: Administrative<br />**3**: Risky<br />**4**: VPN<br />**5**: Cloud provider<br />**6**: Other |
-| subnets | An array of masks as strings (IPv4 / IPv6) |
-| organization (Optional) | The registered ISP |
-| tags (Optional) | An array of new or existing objects including the tag name, id, description, name template, and tenant id |
+| Parameter | Type | Description |
+| --- | --- | --- |
+| name | string | The unique name of the range |
+| _id (Optional)| string | The IP range id. Required if you are not changing the name. |
```
Comment (dagoltz) [inline CloudAppSecurityDocs/api-data-enrichment-update.md:24]: remove the _id parameter as it is used from the URL only

File: CloudAppSecurityDocs/api-data-enrichment-update.md [hunk 3]
```diff
@@ -18,13 +18,17 @@ POST /api/v1/subnet/<ip_range_id>/update_rule/
...
+| category | int | The id of the range category. Providing a category helps you easily recognize activities from interesting IP addresses. Possible values include:<br /><br />**1**: Corporate<br />**2**: Administrative<br />**3**: Risky<br />**4**: VPN<br />**5**: Cloud provider<br />**6**: Other |
+| subnets | list | An array of masks as strings (IPv4 / IPv6) |
+| overrideOrganization (Optional) | bool | Flag to enable or disable overriding the Organization field. |
+| organization (Optional) | string | The registered ISP |
+| tags (Optional) | list | An array of new or existing objects including the tag name, id, description, name template, and tenant id |
+| overrideLocation (Optional) | bool | Flag to enable or disable overriding the Location field. |
+| location (Optional) | string |An object including the location name, latitude, longitude, country code, and country name|
```
Comment (dagoltz) [inline CloudAppSecurityDocs/api-data-enrichment-update.md:31]: the "location" property is a nullable string property that is a stringified version of the location object.
here is an example:
```
"{\"name\":\"Washington\",\"latitude\":47.411376953125,\"longitude\":-120.5563735961914,\"countryCode\":\"US\",\"countryName\":\"United States\",\"regionCode\":\"WA\"}"
```
I would advise to leave it out too if not required as it is a confusiong option

File: CloudAppSecurityDocs/api-data-enrichment-update.md [hunk 4]
```diff
@@ -18,13 +18,17 @@ POST /api/v1/subnet/<ip_range_id>/update_rule/
...
-| subnets | An array of masks as strings (IPv4 / IPv6) |
-| organization (Optional) | The registered ISP |
-| tags (Optional) | An array of new or existing objects including the tag name, id, description, name template, and tenant id |
+| Parameter | Type | Description |
+| --- | --- | --- |
+| name | string | The unique name of the range |
+| _id (Optional)| string | The IP range id. Required if you are not changing the name. |
+| category | int | The id of the range category. Providing a category helps you easily recognize activities from interesting IP addresses. Possible values include:<br /><br />**1**: Corporate<br />**2**: Administrative<br />**3**: Risky<br />**4**: VPN<br />**5**: Cloud provider<br />**6**: Other |
+| subnets | list | An array of masks as strings (IPv4 / IPv6) |
+| overrideOrganization (Optional) | bool | Flag to enable or disable overriding the Organization field. |
```
Comment (dagoltz) [inline CloudAppSecurityDocs/api-data-enrichment-update.md:27]: overrideOrganization is not needed - we use the organization property only (that is nullable)

**Review by dagoltz — CHANGES_REQUESTED** [review]
Hi, I have added comments to the PR - please go over them.

Daniel from MCAS

File: CloudAppSecurityDocs/api-data-enrichment-update.md (hunk 2 above)
Comment (dcurwin) [inline CloudAppSecurityDocs/api-data-enrichment-update.md:24]:
```suggestion

```

File: CloudAppSecurityDocs/api-data-enrichment-update.md (hunk 4 above)
Comment (dcurwin) [inline CloudAppSecurityDocs/api-data-enrichment-update.md:27]:
```suggestion
| overrideOrganization (Optional) | bool | Flag to enable or disable overriding the Organization field. |
```

File: CloudAppSecurityDocs/api-data-enrichment-update.md (hunk 1 above)
Comment (dcurwin) [inline CloudAppSecurityDocs/api-data-enrichment-update.md:30]:
```suggestion

```

File: CloudAppSecurityDocs/api-data-enrichment-update.md (hunk 3 above)
Comment (dcurwin) [inline CloudAppSecurityDocs/api-data-enrichment-update.md:31]:
```suggestion

```

**Comment by PRMerger14** [comment]
@dcurwin : Thanks for your contribution! The author(s) have been notified to review your proposed change.

File: CloudAppSecurityDocs/api-data-enrichment-update.md [hunk 5]
```diff
@@ -18,13 +18,14 @@ POST /api/v1/subnet/<ip_range_id>/update_rule/
...
-| category | The id of the range category. Providing a category helps you easily recognize activities from interesting IP addresses. Possible values include:<br /><br />**1**: Corporate<br />**2**: Administrative<br />**3**: Risky<br />**4**: VPN<br />**5**: Cloud provider<br />**6**: Other |
-| subnets | An array of masks as strings (IPv4 / IPv6) |
-| organization (Optional) | The registered ISP |
-| tags (Optional) | An array of new or existing objects including the tag name, id, description, name template, and tenant id |
+| Parameter | Type | Description |
+| --- | --- | --- |
+| name | string | The unique name of the range |
+| category | int | The id of the range category. Providing a category helps you easily recognize activities from interesting IP addresses. Possible values include:<br /><br />**1**: Corporate<br />**2**: Administrative<br />**3**: Risky<br />**4**: VPN<br />**5**: Cloud provider<br />**6**: Other |
+| subnets | list | An array of masks as strings (IPv4 / IPv6) |
+| overrideOrganization (Optional) | bool | Flag to enable or disable overriding the Organization field. |
```
Comment (dcurwin) [inline CloudAppSecurityDocs/api-data-enrichment-update.md:26]:
```suggestion

```

**Comment by PRMerger3** [comment]
@dcurwin : Thanks for your contribution! The author(s) have been notified to review your proposed change.

**Comment by PRMerger3** [comment]
@dcurwin : Thanks for your contribution! The author(s) have been notified to review your proposed change.

**Review by dcurwin — APPROVED** [review]

**Comment by dcurwin** [comment]
@mcvic1rj - Thank you for your contribution. Based on feedback from the engineering team, we didn't include the additional parameters, but we did include the type column.

File: CloudAppSecurityDocs/api-data-enrichment-update.md (hunk 2 above)
Comment (mcvic1rj) [reply CloudAppSecurityDocs/api-data-enrichment-update.md]: Hey @dagoltz,  this does not appear to be the case for update calls where you are not trying to update the name. Please see the below sample calls/returns.

All calls are POST requests to `https://tenantname.us3.portal.cloudappsecurity.com/api/v1/subnet/<ip_range_id>/update_rule/` with valid credentials.

Case 1, Update subnets without specifying the `name` parameter in the body:
```json
{
  "category":5,
  "subnets":[
    "192.168.1.0/24",
    "192.168.2.0/16"
  ]
}
```
Result:
```json
{
    "errors": [
        {
            "error": "Missing name for the subnet range rule.",
            "errorMessageCode": "CONSOLE_IP_RULE_ERROR_MISSING_NAME_FOR_THE_SUBNET_RULE"
        }
    ]
}
```

Case 2, Update subnets with specifying the `name` parameter in the body:
```json
{
  "name":"testRange",
  "category":5,
  "subnets":[
    "192.168.1.0/24",
    "192.168.2.0/16"
  ]
}
```
Result:
```json
{
    "errors": [
        {
            "error": "A range with the same name already exists.",
            "errorMessageCode": "CONSOLE_IP_RULE_ERROR_DUPLICATE_NAME_FOR_THE_SUBNET_RULE"
        }
    ]
}
```

Case 3, Update subnets with specifying the `name` parameter and `_id` in the body:
```json
{
  "name":"testRange",
  "_id":"<ip_range_id>",
  "category":5,
  "subnets":[
    "192.168.1.0/24",
    "192.168.2.0/16"
  ]
}
```
Result:
```json
{
    "name": "testRange",
    "subnets": [
        {
            "mask": 120,
            "address": "0000:0000:0000:0000:0000:ffff:c0a8:0100",
            "originalString": "192.168.1.0/24"
        },
        {
            "mask": 112,
            "address": "0000:0000:0000:0000:0000:ffff:c0a8:0000",
            "originalString": "192.168.2.0/16"
        }
    ],
    "location": null,
    "organization": null,
    "tags": [],
    "category": 5,
    "lastModified": 1633375827168.9678
}
```

Case 4, Update subnets and `name` without specifying the `_id` parameter in the body:
```json
{
  "name":"testRange2",
  "category":5,
  "subnets":[
    "192.168.1.0/24",
    "192.168.2.0/16"
  ]
}
```
Result:
```json
{
    "name": "testRange2",
    "subnets": [
        {
            "mask": 120,
            "address": "0000:0000:0000:0000:0000:ffff:c0a8:0100",
            "originalString": "192.168.1.0/24"
        },
        {
            "mask": 112,
            "address": "0000:0000:0000:0000:0000:ffff:c0a8:0000",
            "originalString": "192.168.2.0/16"
        }
    ],
    "location": null,
    "organization": null,
    "tags": [],
    "category": 5,
    "lastModified": 1633376093881.685
}
```

**Comment by mcvic1rj** [comment]
> @mcvic1rj - Thank you for your contribution. Based on feedback from the engineering team, we didn't include the additional parameters, but we did include the type column.

Thank you for getting some of the changes included. I have responded to one of @dagoltz comments. Based on the suggestion, it seems like there is either a bug in how the API is processing the request or the `_id` parameter should be included.

File: CloudAppSecurityDocs/api-data-enrichment-update.md (hunk 2 above)
Comment (dagoltz) [reply CloudAppSecurityDocs/api-data-enrichment-update.md]: Hi @mcvic1rj ,
the API is built to accept all fields at all time, it is not built for partial updates.
so as the name property is required, any update without it (even if you add the not-used "_id" property) will be responded with an error.

partial updates will usually be done using PATH or PUT HTTP verbs, which we do not support for this endpoint.

**Pull request merged**